service/src/main/resources/api/service_openapi.yaml

```diff
@@ -994,6 +1093,129 @@ components:
+          # from the UI to the backend.
+          $ref: "#/components/schemas/FilterV2"
+
+    CohortUpdateInfoV2:
+      type: object
+      properties:
+        displayName:
+          $ref: "#/components/schemas/CohortDisplayNameV2"
+        description:
+          $ref: "#/components/schemas/CohortDescriptionV2"
+        criteriaGroups:
+          type: array
+          description: Groups of criteria that define the cohort
+          items:
+            $ref: "#/components/schemas/CriteriaGroupV2"
+        filter:
+          # TODO: Remove this filter property once we move the code that converts the criteriaGroups to a filter
+          # from the UI to the backend.
+          $ref: "#/components/schemas/FilterV2"
+
+    UnderlayNameV2:
+      type: string
+      description: Name of the underlay, immutable
+
+    CohortDisplayNameV2:
+      type: string
+      description: Human readable name of the cohort
+
+    CohortDescriptionV2:
+      type: string
+      description: Description of the cohort
+
+    CriteriaGroupV2:
+      type: object
+      description: Group of criteria for a cohort or concept set definition
+      properties:
+        id:
+          type: string
+          description: ID of the group, immutable
+        displayName:
+          type: string
+          description: Name of the group
+        criteria:
+          type: array
+          description: Set of criteria in the group
+          items:
+            $ref: "#/components/schemas/CriteriaV2"
+        operator:
+          type: string
+          description: Operator to use when combining the criteria in the group
+          enum: ["AND", "OR"]
+        excluded:
+          type: boolean
+          description: True to exclude the group, false to include it
+
+    CriteriaV2:
+      type: object
+      description: Single criteria for a cohort or concept set definition
+      properties:
+        id:
+          type: string
+          description: ID of the criteria, immutable
+        displayName:
+          type: string
+          description: Name of the criteria
+        uiPluginName:
+          type: string
+          description: Name of the UI plugin that generated this criteria
+        uiData:
+          type: string
+          description: UI representation of the criteria, includes the user's selection, plugin-specific
+          format: byte # Base-64 encoded string.
+        uiConfig:
+          # TODO: Remove this uiConfig property once we fully support UI plugins, because it should be identical
+          # to the static config stored in the underlay. For now, we save it separately for each criteria so we
+          # can handle schema changes in the UI code instead of in plugin-specific backend code.
+          type: string
+          description: UI configuration for the criteria, plugin-specific
+          format: byte # Base-64 encoded string.
```

Comment:
This isn't a very practical format to work with. Is the expectation that these versions are for the API only and that there will be separate versions for the UI/backend/elsewhere?

I also don't understand the format. Isn't it a just a regular string when it's base64? Does the Java OpenAPI plugin not support the generic `object` type so this was necessary instead?

Reply:
Yeah having an `object` without any specific properties defined doesn't really work in Java. Swagger generates Java classes based on the OpenAPI spec, but if there's no properties, we just get an empty object with no information.

RE format: Would a regular string type work better for the UI? I don't think it matters on the backend, because I'm just going to store it as is, without parsing it.

Can you clarify what you mean about the separate versions for the UI/backend/elsewhere? My understanding of how we want this to work in the short-term is:
- User makes selection. UI stores the selection information in the `selectionData` property, and the static UI config for this criteria in the `uiConfig` property.
- UI saves the cohort, passing both properties to the backend. Backend saves both the `selectionData` and `uiConfig` in its internal DB.
- Another day, UI fetches the cohort from the backend. Backend pulls both the `selectionData` and `uiConfig` from its internal DB and returns them to the UI. The UI checks the `uiConfig` to see if it's out of date with the current static UI config. If so, it handles the version differences itself.

And in the long-term:
- User makes selection. UI stores the selection information in the `selectionData` property.
- UI saves the cohort, passing only the `selectionData` property to the backend. Backend asks the appropriate plugin what the current static `uiConfig` is, and stores both that and the `selectionData` in its internal DB.
- Another day, UI fetches the cohort from the backend. Backend pulls both the `selectionData` and `uiConfig` from its internal DB and passes them to the appropriate plugin. The plugin checks the `uiConfig` to see if it's out of date with the current static UI config. If so, it upgrades the `selectionData` to work with the current version. The backend passes this (possibly upgraded) `selectionData` to the UI.

Does this match your thinking, or what am I missing?

Reply:
The UI (or plugin) has to work with the actual data contained in uiConfig and uiData. If they're stored as strings then they have to be unpacked to be used while rendering or making edits. I don't want to be constantly unpacking them so I'll need a different Criteria object for a working copy.

In any case, there's no reason to store them in base64. They'll contain JSON not binary data.

Reply:
Per offline conversation, we need to figure out how to handle plugin-specific schemas. Plan to tackle this work as part of fully supporting UI plugins on the backend. The UI needs the schema so it can pass the objects to plugins. The backend will also need the schema for both the `selectionData` and `uiConfig` so it can deserialize the objects in its internal DB to pass to the plugin Java code.

One approach may be to have plugins define these schemas in a separate OpenAPI file and point Tanagra at it. Then the backend and UI can both generate the appropriate classes/objects from this schema, but we don't have to modify our service API definition for every new plugin. We could probably also inject custom schemas into our service API definition, but that seems more unstable to me.

Anyway, for now we're going to pass the serialized representation between the UI to the backend as a string. The UI will have to deserialize it, and the backend is just going to store it as a blob.